### PR TITLE
Fixes for attach-detach controller enablement on existing nodes

### PR DESCRIPTION
**Pull request by pmorie**
Upstream patches for:

1.  kubernetes/kubernetes#31531
2.  kubernetes/kubernetes#31730

Bug 1370312

### CODE REVIEW

**Comment by pmorie** [comment]
[test]

**Comment by pmorie** [comment]
@liggitt want to give this a look-see and ensure I haven't destroyed the patch we carried from #6310?

**Comment by smarterclayton** [comment]
I want to know whether the upstream patch ever got applied.

**Comment by eparis** [comment]
it was LGTM, but flake in unit test twice.

**Comment by pmorie** [comment]
@smarterclayton the upstream is LGTM, in the submit queue now.

[test]

File: vendor/k8s.io/kubernetes/pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1148,78 +1151,149 @@ func (kl *Kubelet) providerRequiresNetworkingConfiguration() bool {
...
+				glog.Errorf("Unable to reconcile node %q with API server: error updating node: %v", kl.nodeName, err)
+				return false
 			}
-			if currentNode.Spec.ExternalID == node.Spec.ExternalID {
+		}
+
+		return true
+	}
+
+	if kl.cloud == nil {
```
Comment (liggitt) [inline vendor/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: this is a carry we have (https://github.com/openshift/origin/commit/88abe47b9a963b41acd0c4f0fd18827192648468). We should keep the carry distinct from the upstream pick commit

**Comment by liggitt** [comment]
does this primarily impact nodes upgraded from previous releases? have sufficient upgrade tests been run upstream? trying to understand the potential risk and impact

File: vendor/k8s.io/kubernetes/pkg/kubelet/kubelet.go (hunk 1 above)
Comment (pmorie) [inline vendor/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: @liggitt I'm cool with doing it that way.

**Comment by pmorie** [comment]
@liggitt yes, it fixes an issue where a node upgraded from a previous release that has the configuration item set to enable controller attach/detach will not annotate the node resource to tell the controller to do the attach.  So, the node will wait for the controller to do the attach, and the controller never actually does it, which breaks all attachable volumes (basically every PV type except hostpath and nfs).

File: vendor/k8s.io/kubernetes/pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1228,6 +1228,31 @@ func (kl *Kubelet) tryRegisterWithApiServer(node *api.Node) bool {
...
+				requiresUpdate := kl.reconcileCMADAnnotationWithExistingNode(node, existingNode)
+				if requiresUpdate {
+					if _, err := kl.kubeClient.Core().Nodes().Update(existingNode); err != nil {
+						glog.Errorf("Unable to reconcile node %q with API server: error updating node: %v", kl.nodeName, err)
+						return false
+					}
+				}
+			}
+
+			return true
```
Comment (liggitt) [inline vendor/k8s.io/kubernetes/pkg/kubelet/kubelet.go]: move inside the `existingNode.Spec.ExternalID == addr.Address` block

**Comment by liggitt** [comment]
LGTM

**Comment by pmorie** [comment]
@liggitt tag and bag? :art: :pouch:

**Comment by liggitt** [comment]
[merge]

**Pull request merged**